google-cloud-firestore/lib/google/cloud/firestore/query.rb

```diff
@@ -0,0 +1,613 @@
+      #
+      # Instances of this class are immutable. All methods that refine the query
+      # return new instances.
+      #
+      # @example
+      #   require "google/cloud/firestore"
+      #
+      #   firestore = Google::Cloud::Firestore.new
+      #
+      #   # Create a query
+      #   query = firestore.col(:cities).select(:population)
+      #
+      #   query.get do |city|
+      #     puts "#{city.document_id} has #{city[:population]} residents."
+      #   end
+      #
+      class Query
+        ##
+        # @private The parent path for the query.
+        attr_accessor :parent_path
+
+        ##
+        # @private The Google::Firestore::V1beta1::Query object.
+        attr_accessor :query
+
+        ##
+        # @private The firestore client object.
+        attr_accessor :client
+
+        ##
+        # Restricts documents matching the query to return only data for the
+        # provided fields.
+        #
+        # @param [String, Symbol, Array<String|Symbol>] fields A field mask to
+        #   filter results with and return only the specified fields. One or
+        #   more field paths can be specified.
+        #
+        # @return [Query] New query with `select` called on it.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.select(:population)
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def select *fields
+          new_query = @query.dup
+          new_query ||= StructuredQuery.new
+
+          fields = fields.flatten.map(&:to_s).reject(&:nil?).reject(&:empty?)
+          new_query.select ||= StructuredQuery::Projection.new
+          fields.each do |field|
+            field = StructuredQuery::FieldReference.new field_path: field.to_s
+            new_query.select.fields << field
+          end
+
+          Query.start new_query, parent_path, client
+        end
+
+        ##
+        # Selects documents from all collections, immediate children and nested,
+        # of where the query was created from.
+        #
+        # @return [Query] New query with `all_descendants` called on it.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.all_descendants
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def all_descendants
```

Comment:
Sorry, this is an instance of where I am going to ask you to remove this functionality purely because it is not yet available in any of the other SDKs. We would like to figure out how we can expose it in the mobile SDKs first.

Please keep this somewhere (you could even just mark in private if Ruby enforces it), so that we can resurrect this soon.

Reply:
I would remove these, as we have removed the other items you have requested, except that we depend on this functionality to clean up after our acceptance tests. So I'm afraid I will have to keep these methods and mark these as private so they won't appear in the documentation.